src/modules/Elsa.Workflows.Api/Requirements/ReadOnlyRequirement.cs

```diff
@@ -0,0 +1,64 @@
+﻿using Elsa.Common.Models;
+using Elsa.Workflows.Api.Options;
+using Elsa.Workflows.Management.Contracts;
+using Elsa.Workflows.Management.Filters;
+using JetBrains.Annotations;
+using Microsoft.AspNetCore.Authorization;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.Options;
+
+namespace Elsa.Workflows.Api.Requirements;
+
+public class ReadOnlyRequirement : IAuthorizationRequirement
+{
+}
+
+
+/// <inheritdoc />
+[PublicAPI]
+public class ReadOnlyRequirementHandler : AuthorizationHandler<ReadOnlyRequirement>
+{
+    private readonly IOptions<ApiOptions> _apiOptions;
+    private readonly IWorkflowDefinitionStore _store;
+    private readonly IHttpContextAccessor _httpContextAccessor;
+
+    /// <inheritdoc />
+    public ReadOnlyRequirementHandler(
+        IOptions<ApiOptions> apiOptions,
+        IWorkflowDefinitionStore store,
+        IHttpContextAccessor httpContextAccessor)
+    {
+        _apiOptions = apiOptions;
+        _store = store;
+        _httpContextAccessor = httpContextAccessor;
+    }
+
+    /// <inheritdoc />
+    protected override async Task HandleRequirementAsync(AuthorizationHandlerContext context, ReadOnlyRequirement requirement)
+    {
+        if (_apiOptions.Value.IsReadOnlyMode)
+        {
+            context.Fail(new AuthorizationFailureReason(this, "Workflow edit is not allowed when the read-only mode is enabled."));
+        }
+
+        var definitionId = _httpContextAccessor.HttpContext?.Request.RouteValues["definitionId"]?.ToString();
```

Comment:
I noticed that the context contains the Resource property that has access to some things, including the http context.

Are you saying to use that ne instead of the IHttpContextAccessor? Did I understand correctly?

Reply:
I think that the Resource object is set by whomever is calling IAuthorizationService.AuthorizeAsync. Possibly, FastEndpoints stashes some HTTP related things in there, since it presumable executes IAuthorizationService.

But when you invoke IAuthorizationService yourself, you can provide the workflow definition object as the resource to which access is being authorized.

Here's more information about this approach: https://learn.microsoft.com/en-us/aspnet/core/security/authorization/resourcebased?view=aspnetcore-8.0